File: ExtremeParser/Parsers/Defines.vb
Imports System.Text.RegularExpressions

Public Class Defines
    Public Shared Sub Parse(ByRef code As String, fileName As String, ByRef parts As CodeParts, Optional add As Boolean = True)
        For Each Match As Match In Regex.Matches(code, "^[ \t]*[#]define[ \t]+(?<name>[^\s\\;]+)[ \t]*(?:\\\s+)?(?>(?<value>[^\\\n\r]+)[ \t]*(?:\\\s+)?)*", RegexOptions.Multiline)
            Dim defineName As String = Match.Groups(1).Value
            Dim defineValue As String = ""

            For Each capt As Capture In Match.Groups(2).Captures
                defineValue += capt.Value.Trim + vbCrLf
            Next

            Try
                If defineName.Contains("%") Then
                    If add Then
                        parts.Macros.Insert(0, New DefinesStruct(defineName.Trim, defineValue.Trim))
                    Else
                        parts.Macros.RemoveAll(Function(x) x.DefineName = defineName)
                    End If
                Else
                    If add Then
                        parts.Defines.Insert(0, New DefinesStruct(defineName.Trim, defineValue.Trim))
                    Else
                        parts.Defines.RemoveAll(Function(x) x.DefineName = defineName)
                    End If
                End If

            Catch ex As Exception
            End Try
        Next
    End Sub
End Class

File: ExtremeParser/Parsers/Defines.cs
using System;
using System.Linq;
using System.Text.RegularExpressions;
using ExtremeParser.Types;

namespace ExtremeParser.Parsers
{
    public class Defines
    {
        public static void Parse(string code, string fileName, CodeParts parts, bool add = true)
        {
            foreach (Match match in Regex.Matches(code, "^[ \\t]*[#]define[ \\t]+(?<name>[^\\s\\\\;]+)[ \\t]*(?:\\\\\\s+)?(?>(?<value>[^\\\\\\n\\r]+)[ \\t]*(?:\\\\\\s+)?)*", RegexOptions.Multiline))
            {
                string defineName = Convert.ToString(match.Groups[1].Value);
                string defineValue = match.Groups[2].Captures.Cast<Capture>().Aggregate("", (current, capt) => current + (capt.Value.Trim() + "\r\n"));
			
                try
                {
                    if (defineName.Contains("%"))
                    {
                        if (add)
                        {
                            parts.Macros.Insert(0, new DefinesStruct(defineName.Trim(), defineValue.Trim()));
                        }
                        else
                        {
                            parts.Macros.RemoveAll(x => x.DefineName == defineName);
                        }
                    }
                    else
                    {
                        if (add)
                        {
                            parts.Defines.Insert(0, new DefinesStruct(defineName.Trim(), defineValue.Trim()));
                        }
                        else
                        {
                            parts.Defines.RemoveAll(x => x.DefineName == defineName);
                        }
                    }
				
                }
                catch (Exception)
                {
                }
            }
        }
    }
}
